### PR TITLE
Vue: Improve `v-for` format

### LINKED ISSUE #8769: Vue: weird `v-for` format
**Prettier 2.0.5**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAeeBbADgGwIbwB8AOlAARmo4CWAzjKRRQG4C0AZhAE4C8xIjJmQAUgoeIrAxEmWRzQA5gBlFABS4QsAGjmKVUBes3TZFAL5aTs6lAAmcAB5WAlGQjtdB-cr2+vfny86BnImfkESUNQAehp6SJjMXAI4SJAtEE0YamhaZFA8Lg0Ad1VChDyUPBxivABPPIyAIy48MABrOBgAZSw2mwVkGC4AVzgMgAsYDBwAdQnqeFo+sDhuisXqZkW65HBaRpAbWjguGHU8BQw8ZHZqk4yAK1oHACFWjq7uvAw4JRs4Ld7uMQM8HN0Bjg4ABFEYQeBAnAPEB9LgnLh7ZhjdIorg2GCzai2GATZAADgADBksBoTrNWlg9jS4OjmICMgBHOHwIxYSogPC0VhQOBwey2HFcOBc6hSi5XG5IO5IkEnDDUIajVWQmHcwFK4EZGB4JqE4mkpAAJiNrWoNAMAGEIBhrnsWQBWHEjE4AFRNlWVyKxcAAknYED0wHisDAAIJ2bowOpQxEnMxmIA)
```sh
--parser vue
```

**Input:**
```vue
<template>
  <list
    v-for="
     (
             {
                 longLongProp, longLongProp
              },
             index
      ) of longLongLongLongLongLongLongLongList
    "
  >
  </list>
</template>

```

**Output:**
```vue
<template>
  <list
    v-for="({ longLongProp, longLongProp },
    index) of longLongLongLongLongLongLongLongList"
  >
  </list>
</template>

```

**Expected behavior:**

Something nicer, maybe break around `of`

### PR DESCRIPTION
**Pull request by zweimach**
I managed to get this:

```vue
<template>
  <list
    v-for="({ longLongProp, longLongProp }, index)
    of longLongLongLongLongLongLongLongList"
  >
  </list>
</template>
```
and another problem arises:
```vue
<template>
  <list
    v-for="({
      longLongProp,
      longLongProp,
      anotherLongLongProp,
      yetAnotherLongLongProp,
    },
    index)
    of longLongLongLongLongLongLongLongList"
  >
  </list>
</template>
```

Hopefully, this will close #8769.

- [x] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory)
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/pr-XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

File: tests/vue/vue/__snapshots__/jsfmt.spec.js.snap [hunk 1]
```diff
@@ -52,6 +56,25 @@ semi: false
...
+    },
+    index)
+    of longLongLongLongLongLongLongLongList"
+    v-for="{
+      longLongProp,
+      longLongProp,
+      anotherLongLongProp,
+      yetAnotherLongLongProp,
+    }
+    of longLongLongLongLongLongLongLongList"
```
Comment (fisker) [inline tests/vue/vue/__snapshots__/jsfmt.spec.js.snap:77]: Still ugly 😢

File: src/language-html/syntax-vue.js [hunk 2]
```diff
@@ -14,12 +14,16 @@ function printVueFor(value, textToDoc) {
...
-      })
+      concat([
+        group(
+          textToDoc(`function _(${left}) {}`, {
+            parser: "babel",
+            __isVueForBindingLeft: true,
+          })
+        ),
+        line,
+      ])
```
Comment (fisker) [inline src/language-html/syntax-vue.js:25]: I didn't know we do this trick, why don't we use

```js
textToDoc(`for (${value}) {}`)
```

Comment (fisker) [reply src/language-html/syntax-vue.js:25]: Sorry, bad idea.

**Comment by zweimach** [comment]
@fisker what do you think?

```vue
<template>
  <list
    v-for="({ longLongProp, longLongProp }, index)
    of longLongLongLongLongLongLongLongList"

    v-for="{
      longLongProp,
      longLongProp,
    } of longLongLongLongLongLongLongLongList"

    v-for="({
      longLongProp,
      longLongProp,
      anotherLongLongProp,
      yetAnotherLongLongProp,
    },
    index)
    of longLongLongLongLongLongLongLongList"
  >
  </list>
</template>
```

**Comment by fisker** [comment]
I think the second one is good, the first one not very pretty but acceptable, the third one is very ugly.

**Comment by zweimach** [comment]
> I think the second one is good, the first one not very pretty but acceptable, the third one is very ugly.

What would be good for the first and third cases?

**Comment by fisker** [comment]
Like this?

```vue
<template>
  <list
    v-for="(
      { longLongProp, longLongProp }, index
    ) of longLongLongLongLongLongLongLongList"

    v-for="{
      longLongProp,
      longLongProp,
    } of longLongLongLongLongLongLongLongList"

    v-for="(
      {
        longLongProp,
        longLongProp,
        anotherLongLongProp,
        yetAnotherLongLongProp,
      }, index
    ) of longLongLongLongLongLongLongLongList"
  >
  </list>
</template>
```

**Comment by zweimach** [comment]
Consider it done. But I get this

```vue
<template>
  <list
    v-for="(
      {
        longLongProp,
        longLongProp,
        anotherLongLongProp,
        yetAnotherLongLongProp,
      },
      index
    ) of longLongLongLongLongLongLongLongList"
  >
  </list>
</template>
```

instead of

```vue
<template>
  <list
    v-for="(
      {
        longLongProp,
        longLongProp,
        anotherLongLongProp,
        yetAnotherLongLongProp,
      }, index // How can I force this one?
    ) of longLongLongLongLongLongLongLongList"
  >
  </list>
</template>
```

**Comment by zweimach** [comment]
Should I also make `changelog_unreleased/*/pr-XXXX.md`?

**Comment by fisker** [comment]
Looks good already.

File: src/language-js/html-binding.js [hunk 3]
```diff
@@ -19,12 +19,21 @@ function printHtmlBinding(path, options, print) {
...
+              group(
+                join(
+                  concat([",", line]),
+                  functionDeclarationPath.map(print, "params")
+                )
+              ),
+            ])
           ),
+          params.length > 1 ? softline : "",
           params.length > 1 ? ")" : "",
```
Comment (fisker) [inline src/language-js/html-binding.js:37]: There are many `params.length > 1`, how about a `if/else`.

File: tests/vue/vue/__snapshots__/jsfmt.spec.js.snap [hunk 4]
```diff
@@ -52,6 +59,47 @@ semi: false
...
+      longLongProp,
+      anotherLongLongProp,
+      yetAnotherLongLongProp,
+    } of longLongLongLongLongLongLongLongList"
+    v-for="(
+      objectPropKey, { firstValue, secondValue }, index
+    ) in objectWithAVeryLongLongLongLongName"
+    v-for="(
+      objectPropKey, { firstValue, secondValue, thirdValue, fourthValue }, index
+    ) in objectWithAVeryLongLongLongLongName"
```
Comment (fisker) [inline tests/vue/vue/__snapshots__/jsfmt.spec.js.snap:89]: Is this a valid case for vue?

**Review by fisker — COMMENTED** [review]
Overall looks very good

**Comment by fisker** [comment]
> Should I also add these changes to changelog_unreleased/*/pr-XXXX.md?

Yes, please.

File: tests/vue/vue/__snapshots__/jsfmt.spec.js.snap (hunk 4 above)
Comment (zweimach) [reply tests/vue/vue/__snapshots__/jsfmt.spec.js.snap:89]: Whoopsie, my bad, I accidentally swapped the first and second params.

Comment (fisker) [reply tests/vue/vue/__snapshots__/jsfmt.spec.js.snap:89]: Oh, it is `value, key, index`, never use this syntax 👍

**Comment by fisker** [comment]
Please add some array destructuring tests, and some nest destructuring.

File: src/language-js/html-binding.js (hunk 3 above)
Comment (zweimach) [reply src/language-js/html-binding.js:37]: Good enough?

```js
      (functionDeclarationPath) => {
        const { params } = functionDeclarationPath.getValue();
        let [blankOrSoftline, indentOrGroup] = ["", group];
        let [openParen, closeParen] = ["", ""];
        if (params.length > 1) {
          // @ts-ignore
          [blankOrSoftline, indentOrGroup] = [softline, indent];
          [openParen, closeParen] = ["(", ")"];
        }
        return concat([
          openParen,
          indentOrGroup(
            concat([
              blankOrSoftline,
              group(
                join(
                  concat([",", line]),
                  functionDeclarationPath.map(print, "params")
                )
              ),
            ])
          ),
          blankOrSoftline,
          closeParen,
        ]);
      },
```

Comment (fisker) [reply src/language-js/html-binding.js:37]: Why not just

```
if (params.length > 1) {
return ...
}

return ...
```

Comment (zweimach) [reply src/language-js/html-binding.js:37]: Fair enough.

**Review by fisker — APPROVED** [review]
Good job, thanks.

**Comment by zweimach** [comment]
Thank you for your guidance.

**Review by evilebottnawi — APPROVED** [review]

**Review by sosukesuzuki — APPROVED** [review]
Looks good, thanks!

**Comment by Shinigami92** [comment]
When can we expect a next release? At least a `2.1.3`?

**Pull request merged**